Set nest entities as unavailable on lost connection

PeteRager opened the pull request:
Update Climate and Sensor entities to be unavailable when the device connectivity trait indicates the device is offline.

The prior behavior, the last known values would be displayed indefinitely if the device lost internet connectivity.  This was creating the illusion that the device was still connected.  With this change, the Home Assistant entities will become unavailable when the device loses connectivity.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #70479
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

PeteRager committed:
NEST - Issues with lost internet connectivity #70479

Update Climate and Sensor entities to be unavailable when the device connectivity trait indicates the device is offline.

The prior behavior, the last known values would be displayed indefinitely if the device lost internet connectivity.  This was creating the illusion that the device was still connected.  With this change, the Home Assistant entities will become unavailable when the device loses connectivity.

PeteRager committed:
Update formatting

PeteRager commented:
Initial draft of change, unit tests work as expected.  Next step is to perform system testing on the NEST thermostat by taking it offline then restoring connectivity to verify the changes work as expected with a real device.

PeteRager committed:
Add doc strings, fix indentation

PeteRager committed:
Fix doc strings

PeteRager committed:
Update test_climate_sdm.py

PeteRager committed:
Update test_climate_sdm.py

PeteRager committed:
Update test_sensor_sdm.py

PeteRager committed:
Update test_sensor_sdm.py

PeteRager committed:
more formatting fixes

PeteRager commented:
I have successfully completed functional testing.

Test 1 - with NEST thermostat  in online state, start HA; verify climate, temperature, humidity entities ARE available.  Test Result: Verified
Test 2 - with NEST thermostat in offline state, start HA; verify climate, temperature, humidity entities ARE NOT available. Test Result: Verified
Test 3 - with NEST thermostat in online state, start HA, remove power from NEST thermostat, determine how long it takes for the device to go offline, verify climate, temperature, humidity entities become NOT available. Test Result: 2 hours 3 minutes, verified
Test 4 - with NEST thermostat in offline state from Test 3, reapply power, determine how long it takes for device to go online, verify climate, temperature, humidity entities become available.  Test Result: 2 minutes, verified. 

Summary: change accurately tracks the connectivity trait from the SDM, which is an improvement to reporting stale data indefinitely.  The device remained on WiFi for 2 hours with no power, once it goes offline due to low battery then it becomes marked by google as Offline.  However, the NEST app knew within 2 minutes that the device had lost power - reported as E74.   After 2 hours Nest App declares it offline and the status match.  Net:  google has work to do on the SDM to fully reflect device status.

emontnemery commented on homeassistant/components/nest/climate_sdm.py:
```diff
@@ -112,6 +112,15 @@ def unique_id(self) -> str | None:
+            trait: ConnectivityTrait = self._device.traits[ConnectivityTrait.NAME]
+            if trait.status == CONNECTIVITY_TRAIT_OFFLINE:
+                return False
+        return super().available
```
I don't think we should call `super()` here since `ThermostatEntity` implements its own logic for availability.
```suggestion
        return True
```

PeteRager replied:
Makes sense - updating

emontnemery commented on homeassistant/components/nest/sensor_sdm.py:
```diff
@@ -68,6 +72,15 @@ async def async_added_to_hass(self) -> None:
+            trait: ConnectivityTrait = self._device.traits[ConnectivityTrait.NAME]
+            if trait.status == CONNECTIVITY_TRAIT_OFFLINE:
+                return False
+        return super().available
```
```suggestion
        return True
```

PeteRager replied:
Makes sense updating

emontnemery commented on homeassistant/components/nest/sensor_sdm.py:
```diff
@@ -68,6 +72,15 @@ async def async_added_to_hass(self) -> None:
             self._device.add_update_listener(self.async_write_ha_state)
         )
 
+    @property
+    def available(self) -> bool:
+        """Return entity availability."""
+        if ConnectivityTrait.NAME in self._device.traits:
+            trait: ConnectivityTrait = self._device.traits[ConnectivityTrait.NAME]
+            if trait.status == CONNECTIVITY_TRAIT_OFFLINE:
+                return False
+        return super().available
```
Move this to a mixin-class `NextAvailability` which you extend `SensorBase` and `ThermostatEntity` with.

PeteRager replied:
Created new module availability_mixin.py and class AvailabilityMixin, put logic in here and added to SensorBase and ThermostatEntity

PeteRager replied:
I had a challenge in the mixin making all the linter happy about accessing the _device property from the final class.  I did get something to finally make everyone happy.  If there's a better solve, let me know.

allenporter replied:
One suggestion: There's also a `DeviceInfo` class which is basically a small wrapper for interpreting the nest response APIs. While its generic and used for multiple device types, it seems a little more straight forward to add the shared logic there than accessing attributes of an object this is mixed in with.

PeteRager replied:
> One suggestion: There's also a `DeviceInfo` class which is basically a small wrapper for interpreting the nest response APIs. While its generic and used for multiple device types, it seems a little more straight forward to add the shared logic there than accessing attributes of an object this is mixed in with.

I like it

I'll add it in there and then implement "def available" in those two classes returning self._device.available

emontnemery reviewed (commented):
Looks good, just a couple of minor comments.

PeteRager committed:
Place availability logic in mixin

1. Consolidate repeated code into mixin and apply mixin to Climate and Sensor entities
2. Return true instead of super.available()
3. No unit test changes required to maintain code coverage

PeteRager committed:
Merge branch 'dev' into dev

PeteRager committed:
Define self._device is mixin to make linter happier

PeteRager committed:
Merge branch 'dev' of https://github.com/PeteRager/core into dev

PeteRager committed:
Remove logger used for debugging

PeteRager committed:
restore whitespace

PeteRager committed:
Fix test due to underlying merge change

PeteRager committed:
Update availability_mixin.py

PeteRager commented:
@emontnemery - I've addressed the requested changes.  Apologies for the numerous commits to overcome linter problems; i now have these checks working (i think) locally for the future.

PeteRager committed:
Move availability logic into device_info

PeteRager committed:
Merge branch 'dev' into dev

PeteRager committed:
Update sensor_sdm.py

PeteRager commented:
@emontnemery @allenporter 

I made Allen's requested changes.  Getting a strange error on the hassfest test.  I see that there was just a commit to integrations.json - https://github.com/home-assistant/core/pull/78801

CI / Check hassfest (pull_request) Failing after 43s

General errors:
* [CONFIG_FLOW] File integrations.json is not up to date. Run python3 -m script.hassfest

PeteRager committed:
Merge branch 'dev' into dev
